src/main/java/teammates/storage/api/StudentsDb.java

```diff
@@ -174,18 +209,35 @@ public StudentAttributes getStudentForGoogleId(String courseId, String googleId)
     }
     
     /**
-     * Works for both encrypted keys and unencrypted keys
-     *   (sent out before we started encrypting keys). <br>
+     * Works only for encrypted keys.
+     * 
      * Preconditions: <br>
      * * All parameters are non-null.
      * @return null if no matching student.
      */
     public StudentAttributes getStudentForRegistrationKey(String registrationKey) {
         Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, registrationKey);
+   
         try {
+            // CourseStudent
             String originalKey = StringHelper.decrypt(registrationKey.trim());
+            CourseStudent courseStudent = getCourseStudentEntityForRegistrationKey(originalKey);
+            if (courseStudent != null) {
+                return new StudentAttributes(courseStudent);
+            }
+        } catch (Exception e) {
+            // even if the student does not exist, it's not supposed to throw an exception
+            log.severe("Exception thrown trying to retrieve CourseStudent \n"
+                       + TeammatesException.toStringWithStackTrace(e));
+            // fall back on Student
+        }
+        
+        try {
+            // Student
+            // First, try to retrieve the student by assuming the given registrationKey key is encrypted
```

Comment:
Granted that there may still be students with old registration key, we don't handle them anymore so this comment should not be there anymore.